scala/tempo/src/main/scala/com/databrickslabs/tempo/asofJoin.scala

```diff
@@ -0,0 +1,149 @@
+
+    TSDF(prefixedDF,ts_colName, partitionColstrings:_*)
+  }
+
+  // Add columns from other DF ahead of combining them through Union
+  // TODO: implement unit test
+  def addColumnsFromOtherDF(tsdf: TSDF, other_cols: Seq[StructField]): TSDF = {
+
+    val newDF = other_cols.foldLeft(tsdf.df)((df, colStruct) => df
+      .withColumn(colStruct.name, lit(null)))
+
+    // TODO: fix partition column names, it's not nice this way
+    TSDF(newDF, tsdf.tsColumn.name, tsdf.partitionCols.map(_.name):_*)
+  }
+
+  // union the time series A and time series B data frames
+  def combineTSDF(leftTSDF: TSDF, rightTSDF: TSDF): TSDF = {
+
+    val combinedTsCol = "combined_ts"
+
+    val combinedDF = leftTSDF.df
+      .unionByName(rightTSDF.df)
+      .withColumn(combinedTsCol,
+        coalesce(col(leftTSDF.tsColumn.name), col(rightTSDF.tsColumn.name)))
+
+    TSDF(combinedDF, combinedTsCol, leftTSDF.partitionCols.map(_.name):_*)
+  }
+
+  // helper method to obtain the columns from time series B to paste onto time series A
+  def getLastRightRow(tsdf: TSDF, left_ts_col: StructField, rightCols: Seq[StructField]): TSDF = {
+    // TODO: Add functionality for secondary sort key
+
+    val window_spec: WindowSpec =  Window
+      .partitionBy(tsdf.partitionCols.map(x => col(x.name)):_*)
+      .orderBy(tsdf.tsColumn.name)
+
+    val df = rightCols
+      .foldLeft(tsdf.df)((df, rightCol) =>
+        df.withColumn(rightCol.name, last(df(rightCol.name), true)
+          .over(window_spec)))
+      .filter(col(left_ts_col.name).isNotNull).drop(col(tsdf.tsColumn.name))
+
+    TSDF(df, left_ts_col.name, tsdf.partitionCols.map(_.name):_*)
+  }
+
+  // Create smaller time-based ranges inside of the TSDF partition
+  def getTimePartitions(combinedTSDF: TSDF, tsPartitionVal: Int, fraction: Double): TSDF = {
+
+    val tsColDouble = "ts_col_double"
+    val tsColPartition = "ts_partition"
+    val partitionRemainder = "partition_remainder"
+    val isOriginal = "is_original"
+
+    val partitionDF = combinedTSDF.df
+      .withColumn(tsColDouble, col(combinedTSDF.tsColumn.name).cast("double"))
+      .withColumn(tsColPartition, lit(tsPartitionVal) * (col(tsColDouble) / lit(tsPartitionVal)).cast("integer"))
+      .withColumn(partitionRemainder,(col(tsColDouble) - col(tsColPartition)) / lit(tsPartitionVal))
+      .withColumn(isOriginal, lit(1)).cache()
+
+    // add [1 - fraction] of previous time partition to the next partition.
+    val remainderDF = partitionDF
+      .filter(col(partitionRemainder) >= lit(1 - fraction))
+      .withColumn(tsColPartition, col(tsColPartition) + lit(tsPartitionVal))
+      .withColumn(isOriginal, lit(0))
+
+    val df = partitionDF.union(remainderDF).drop(partitionRemainder, tsColDouble)
+    val newPartitionColNames = combinedTSDF.partitionCols.map(_.name) :+ tsColPartition
+
+    TSDF(df, combinedTSDF.tsColumn.name, newPartitionColNames:_*)
+  }
+
+  // wrapper method for executing the AS OF join based on various parameters
```

Comment:
opp -- I designed a "lookupWhen" function to enable much simpler post-processing requirements to clean up all the columns. It's very annoying code to build, but it's done, feel free to have a look [HERE](https://github.com/databrickslabs/overwatch/blob/a6e7e1267d6792c935d07ca9ee99478cf5637307/src/main/scala/com/databricks/labs/overwatch/utils/asofJoin.scala#L165)

This code basically does an asOfJoin but utilizes the left tsdf as a driver and returns only the left columns, the left ts col and any right cols that were diff in type or name without its prefix which majorly reduces col cleanup after the join.